lib/common/http1client.c

```diff
@@ -237,62 +236,78 @@ static void on_head(h2o_socket_t *sock, const char *err)
         return;
     }
 
-    /* parse response */
-    num_headers = sizeof(headers) / sizeof(headers[0]);
-    rlen = phr_parse_response(sock->input->bytes, sock->input->size, &minor_version, &http_status, &msg, &msg_len, headers,
-                              &num_headers, 0);
-    switch (rlen) {
-    case -1: /* error */
-        on_error_before_head(client, "failed to parse the response");
-        return;
-    case -2: /* incomplete */
-        h2o_timeout_link(client->super.ctx->loop, client->super.ctx->io_timeout, &client->_timeout);
-        return;
-    }
+    {
+        struct phr_header src_headers[MAX_HEADERS];
+        /* parse response */
+        num_headers = MAX_HEADERS;
+        rlen = phr_parse_response(sock->input->bytes, sock->input->size, &minor_version, &http_status, &msg, &msg_len, src_headers,
+                                  &num_headers, 0);
+        switch (rlen) {
+        case -1: /* error */
+            on_error_before_head(client, "failed to parse the response");
+            return;
+        case -2: /* incomplete */
+            h2o_timeout_link(client->super.ctx->loop, client->super.ctx->io_timeout, &client->_timeout);
+            return;
+        }
 
-    /* convert the header names to lowercase */
-    for (i = 0; i != num_headers; ++i)
-        h2o_strtolower((char *)headers[i].name, headers[i].name_len);
+        for (i = 0; i != num_headers; ++i) {
+            const h2o_token_t *token;
+            char *orig_name;
+
+            orig_name = h2o_strdup(NULL, src_headers[i].name, src_headers[i].name_len).base;
```

Comment:
How about creating a memory pool as a local variable of the function (i.e. `h2o_mem_pool_t pool; h2o_mem_init_pool(&pool)` at the top scope of the function), and allocate memory from that pool instead of calling `h2o_strdup` (and `free`) for every header name?

I believe that that would be a good optimization considering the fact that `h2o_mem_alloc_pool` is super-fast for small objects.

We could also allocate `headers` of type `h2o_headers_t` using the pool (instead of `h2o_headers_t *headers[100]`).